File: ios/native/collector/SMMedia.m
//
//  SMMedia.m
//  Supermap
//
//  Created by Shanglong Yang on 2019/5/15.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import "SMMedia.h"

@implementation SMMedia

-(id)init {
    if (self = [super init]) {
        _location = [self getCurrentLocation];
    }
    return self;
}

- (id)initWithName:(NSString *)name {
    if (self = [super init]) {
        _fileName = name;
        
        self.location = [self getCurrentLocation];
    }
    return self;
}

-(Point2D *)getCurrentLocation {
    GPSData* gpsData = [NativeUtil getGPSData];
    Point2D* pt = [[Point2D alloc] initWithX:gpsData.dLongitude Y:gpsData.dLatitude];
    return pt;
}

-(BOOL)setMediaDataset:(Datasource *)datasource datasetName:(NSString *)datasetName {
    self.datasourse = datasource;
    
    DatasetVector* pDatasetVector;
    FieldInfo* fieldInfo;
    int index = (int)[datasource.datasets indexOf:datasetName];
    if (index != -1) {
        pDatasetVector = (DatasetVector*)[datasource.datasets get:index];
//        if(pDatasetVector.datasetType != POINT)
//            return NO;
//        int tag1 = (int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"MediaFileName"];
////        int tag2 = (int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"MediaFileType"];
//        int tag3 = (int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"ModifiedDate"];
//        int tag4 = (int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"MediaFilePaths"];
//        int tag5 = (int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"Description"];
//        int tag6 = (int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"HttpAddress"];
        
        if ((int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"MediaFileName"] == -1) {
            fieldInfo = [[FieldInfo alloc]init];
            fieldInfo.fieldType = FT_TEXT;
            fieldInfo.name = @"MediaFileName";
            [pDatasetVector.fieldInfos add:fieldInfo];
            [fieldInfo dispose];
        }
        
        if ((int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"ModifiedDate"] == -1) {
            fieldInfo = [[FieldInfo alloc]init];
            fieldInfo.fieldType = FT_TEXT;
            fieldInfo.name = @"ModifiedDate";
            [pDatasetVector.fieldInfos add:fieldInfo];
            [fieldInfo dispose];
        }
        
        if ((int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"Description"] == -1) {
            fieldInfo = [[FieldInfo alloc]init];
            fieldInfo.fieldType = FT_TEXT;
            fieldInfo.name = @"Description";
            [pDatasetVector.fieldInfos add:fieldInfo];
            [fieldInfo dispose];
        }
        
        if ((int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"MediaFilePaths"] == -1) {
            fieldInfo = [[FieldInfo alloc]init];
            fieldInfo.fieldType = FT_TEXT;
            fieldInfo.name = @"MediaFilePaths";
            fieldInfo.maxLength = 600;
            [pDatasetVector.fieldInfos add:fieldInfo];
            [fieldInfo dispose];
        }
        
        if ((int)[pDatasetVector.fieldInfos indexOfWithFieldName:@"HttpAddress"] == -1) {
            fieldInfo = [[FieldInfo alloc]init];
            fieldInfo.fieldType = FT_TEXT;
            fieldInfo.name = @"HttpAddress";
            [pDatasetVector.fieldInfos add:fieldInfo];
            [fieldInfo dispose];
        }
        
//        if(tag1==-1 || tag3==-1 || tag4==-1 || tag5==-1)
//            return NO;
//        if([pDatasetVector.fieldInfos get:tag1].fieldType!=FT_TEXT)
//            return NO;
////        if([pDatasetVector.fieldInfos get:tag2].fieldType!=FT_INT16)
////            return NO;
//        if([pDatasetVector.fieldInfos get:tag3].fieldType!=FT_TEXT)
//            return NO;
//        if([pDatasetVector.fieldInfos get:tag4].fieldType!=FT_TEXT)
//            return NO;
//        if([pDatasetVector.fieldInfos get:tag5].fieldType!=FT_TEXT)
//            return NO;
//        if([pDatasetVector.fieldInfos get:tag6].fieldType!=FT_TEXT)
//            return NO;
//        if(pDatasetVector.prjCoordSys.type != PCST_EARTH_LONGITUDE_LATITUDE)
//            return NO;
    } else {
        @try {
            DatasetVectorInfo *vectInfo = [[DatasetVectorInfo alloc]initWithName:datasetName datasetType:CAD];
            pDatasetVector = [datasource.datasets create:vectInfo];
        }
        @catch (NSException *exception) {
            NSLog(@"ERROR { mediaDataset set failed: %@ }",exception);
            return NO;
        }
        
        if (!pDatasetVector) {
            return NO;
        }
        
        fieldInfo = [[FieldInfo alloc]init];
        fieldInfo.fieldType = FT_TEXT;
        fieldInfo.name = @"MediaFileName";
        [pDatasetVector.fieldInfos add:fieldInfo];
        [fieldInfo dispose];
        
        //        fieldInfo = [[FieldInfo alloc]init];
        //        fieldInfo.fieldType = FT_INT16;
        //        fieldInfo.name = @"MediaFileType";
        //        [pDatasetVector.fieldInfos add:fieldInfo];
        
        fieldInfo = [[FieldInfo alloc]init];
        fieldInfo.fieldType = FT_TEXT;
        fieldInfo.name = @"ModifiedDate";
        [pDatasetVector.fieldInfos add:fieldInfo];
        [fieldInfo dispose];
        
        fieldInfo = [[FieldInfo alloc]init];
        fieldInfo.fieldType = FT_TEXT;
        fieldInfo.name = @"Description";
        [pDatasetVector.fieldInfos add:fieldInfo];
        [fieldInfo dispose];
        
        fieldInfo = [[FieldInfo alloc]init];
        fieldInfo.fieldType = FT_TEXT;
        fieldInfo.name = @"MediaFilePaths";
        fieldInfo.maxLength = 600;
        [pDatasetVector.fieldInfos add:fieldInfo];
        [fieldInfo dispose];
        
        fieldInfo = [[FieldInfo alloc]init];
        fieldInfo.fieldType = FT_TEXT;
        fieldInfo.name = @"HttpAddress";
        [pDatasetVector.fieldInfos add:fieldInfo];
        [fieldInfo dispose];

        pDatasetVector.prjCoordSys = [[PrjCoordSys alloc]initWithType:PCST_EARTH_LONGITUDE_LATITUDE];
    }
    
    self.dataset = pDatasetVector;
    
    return YES;
}

-(void)saveLocationDataToDataset
{
    GeoPoint* mGeoPointTem = [[GeoPoint alloc]init];
    GeoStyle* style = [[GeoStyle alloc] init];
    [style setMarkerSize:[[Size2D alloc] initWithWidth:2 Height:2]];
    [style setLineColor:[[Color alloc] initWithR:50 G:240 B:50]];
    [style setMarkerSymbolID:351];
    [style setFillForeColor:[[Color alloc] initWithR:244 G:50 B:50]];
    [mGeoPointTem setStyle:style];
    
    Recordset* mRecordset = [(DatasetVector*)_dataset recordset:NO cursorType:DYNAMIC];
    
    Point2D* pt = [[Point2D alloc]initWithX:_location.x Y:_location.y];
    if ([[SMap singletonInstance].smMapWC.mapControl.map.prjCoordSys type] != PCST_EARTH_LONGITUDE_LATITUDE) {//若投影坐标不是经纬度坐标则进行转换
        Point2Ds *points = [[Point2Ds alloc]init];
        [points add:pt];
        PrjCoordSys *srcPrjCoorSys = [[PrjCoordSys alloc]init];
        [srcPrjCoorSys setType:PCST_EARTH_LONGITUDE_LATITUDE];
        CoordSysTransParameter *param = [[CoordSysTransParameter alloc]init];

        //根据源投影坐标系与目标投影坐标系对坐标点串进行投影转换，结果将直接改变源坐标点串
        [CoordSysTranslator convert:points PrjCoordSys:srcPrjCoorSys PrjCoordSys:[[SMap singletonInstance].smMapWC.mapControl.map prjCoordSys] CoordSysTransParameter:param CoordSysTransMethod:(CoordSysTransMethod)9603];
        pt = [points getItem:0];
    }
    
    [mGeoPointTem setX:pt.x];
    [mGeoPointTem setY:pt.y];
    [mRecordset addNew:mGeoPointTem];
    
    [mRecordset edit];
    [mRecordset moveLast];
    [mRecordset setStringWithName:@"MediaFileName" StringValue:_fileName];
    
//    if([_fileName hasSuffix:@".mp4"] || [_fileName hasSuffix:@".mov"])
//        [ mRecordset setInt16WithName:@"MediaFileType" shortValue:3];
//    else if ([_fileName hasSuffix:@".jpeg"] || [_fileName hasSuffix:@".jpg"])
//        [ mRecordset setInt16WithName:@"MediaFileType" shortValue:1];
//    else if ([_fileName hasSuffix:@".acc"])
//        [ mRecordset setInt16WithName:@"MediaFileType" shortValue:2];
    
//    if ([_mediaType isEqualToString:@"video"]) {
//        [ mRecordset setInt16WithName:@"MediaFileType" shortValue:3];
//    } else if ([_mediaType isEqualToString:@"image"]) {
//        [ mRecordset setInt16WithName:@"MediaFileType" shortValue:1];
//    } else if ([_mediaType isEqualToString:@"audio"]) {
//        [ mRecordset setInt16WithName:@"MediaFileType" shortValue:2];
//    }
//    [mRecordset setStringWithName:@"MediaFileType" StringValue:_mediaType];
    
    [mRecordset update];
    
    [mRecordset dispose];
}

-(BOOL)saveMedia:(NSArray *)filePaths toDictionary:(NSString *)toDictionary addNew:(BOOL)addNew {
    BOOL res = NO;
    if (_paths.count < 0) return res;

    NSFileManager* fileManager = [NSFileManager defaultManager];
    NSError* error = nil;
    
    NSMutableArray* paths = [[NSMutableArray alloc] initWithArray:filePaths];
    NSString* appHomePath = [NSString stringWithFormat:@"%@/%@", NSHomeDirectory(), @"Documents"];
    
    for (int i = 0; i < paths.count; i++) {
        NSString* filePath = [NSString stringWithFormat:@"%@%@", toDictionary, [paths[i] lastPathComponent]];
        if(
           [fileManager fileExistsAtPath:paths[i]] &&
           ![fileManager fileExistsAtPath:filePath]
           ) {
            if ([fileManager copyItemAtPath:paths[i] toPath:filePath error:&error]) {
                res = YES;
                [fileManager removeItemAtPath:paths[i] error:&error];
                paths[i] = [filePath stringByReplacingOccurrencesOfString:appHomePath withString:@""];
            } else {
                res = NO;
            }
        }
    }
    _paths = paths;
    
    if (res && addNew) [self saveLocationDataToDataset];
    return res;
}

-(BOOL)addMediaFiles:(NSArray *)files {
    if (files == nil || files.count == 0) return NO;
    NSMutableArray* addingArr = [[NSMutableArray alloc] init];
    NSFileManager* fileManager = [NSFileManager defaultManager];
    NSMutableArray* paths = [[NSMutableArray alloc] initWithArray:_paths];
    
    for (int i = 0; i < files.count; i++) {
        BOOL exsit = NO;
        if (![fileManager fileExistsAtPath:files[i]]) continue;
        for (int j = 0; j < paths.count; j++) {
            if ([files[i] isEqualToString:paths[j]]) {
                exsit = YES;
                break;
            }
        }
        if (!exsit) {
            [addingArr addObject:files[i]];
        }
    }

    _paths = [paths arrayByAddingObjectsFromArray:addingArr];

    return YES;
}

-(BOOL)deleteMediaFiles:(NSArray *)files {
    if (files == nil || files.count == 0) return NO;
    NSFileManager* fileManager = [NSFileManager defaultManager];
    NSError* error;
    NSMutableArray* paths = [[NSMutableArray alloc] initWithArray:_paths];

    for (int i = 0; i < files.count; i++) {
        if (![fileManager fileExistsAtPath:files[i]]) continue;
        for (int j = 0; j < paths.count; j++) {
            if ([files[i] isEqualToString:_paths[j]] && [fileManager removeItemAtPath:files[i] error:&error]) {
                [paths removeObjectAtIndex:j];
                break;
            }
        }
    }
    _paths = paths;
    return YES;
}

@end

File: android/src/main/java/com/supermap/smNative/collector/SMMedia.java
package com.supermap.smNative.collector;

import android.content.Context;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.net.Uri;
import android.provider.MediaStore;

import com.supermap.data.CoordSysTransMethod;
import com.supermap.data.CoordSysTransParameter;
import com.supermap.data.CoordSysTranslator;
import com.supermap.data.CursorType;
import com.supermap.data.Dataset;
import com.supermap.data.DatasetType;
import com.supermap.data.DatasetVector;
import com.supermap.data.DatasetVectorInfo;
import com.supermap.data.Datasource;
import com.supermap.data.FieldInfo;
import com.supermap.data.FieldInfos;
import com.supermap.data.FieldType;
import com.supermap.data.GeoPoint;
import com.supermap.data.Point2D;
import com.supermap.data.Point2Ds;
import com.supermap.data.PrjCoordSys;
import com.supermap.data.PrjCoordSysType;
import com.supermap.data.Recordset;
import com.supermap.interfaces.mapping.SMap;
import com.supermap.interfaces.utils.SMFileUtil;
import com.supermap.plugin.LocationManagePlugin;

import java.io.File;
import java.util.ArrayList;

/**
 * @Author: shanglongyang
 * Date:        2019/5/17
 * project:     iTablet
 * package:     iTablet
 * class:
 * description:
 */
public class SMMedia {
    private Datasource datasourse;
    private Dataset dataset;
//    private String datasetName;
//    private String mediaType;
//    private String path;
    private ArrayList<String> paths;
    private String fileName;
//    private Bitmap data;
    private Point2D location;

    public Datasource getDatasourse() {
        return datasourse;
    }

    public void setDatasourse(Datasource datasourse) {
        this.datasourse = datasourse;
    }

    public Dataset getDataset() {
        return dataset;
    }

    public void setDataset(Dataset dataset) {
        this.dataset = dataset;
    }

    public ArrayList<String> getPaths() {
        return paths;
    }

    public void setPaths(ArrayList<String> paths) {
        this.paths = paths;
    }

    public String getFileName() {
        return fileName;
    }

    public void setFileName(String fileName) {
        this.fileName = fileName;
    }

    public Point2D getLocation() {
        return location;
    }

    public void setLocation(Point2D location) {
        this.location = location;
    }

    public SMMedia() {
        this.location = getCurrentLocation();
    }

    public SMMedia(String fileName) {
        this.fileName = fileName;

        this.location = getCurrentLocation();
    }

    private Point2D getCurrentLocation() {
        LocationManagePlugin.GPSData gpsDat = SMCollector.getGPSPoint();
        Point2D pt =  new Point2D(gpsDat.dLongitude,gpsDat.dLatitude);

        return pt;
    }

    public boolean setMediaDataset(Datasource datasource, String datasetName) {
        this.datasourse = datasource;

        DatasetVector pDatasetVector;
        FieldInfo fieldInfo;
        int index = datasource.getDatasets().indexOf(datasetName);
        if (index > -1) {
            // Dataset存在
            pDatasetVector = (DatasetVector)datasource.getDatasets().get(index);
            if (pDatasetVector.getType() != DatasetType.POINT && pDatasetVector.getType() != DatasetType.CAD) return false;
            FieldInfos fieldInfos = pDatasetVector.getFieldInfos();
//            int tag1 = fieldInfos.indexOf("MediaFileName");
//            int tag2 = fieldInfos.indexOf("MediaFileType");
//            int tag3 = fieldInfos.indexOf("ModifiedDate");
//            int tag4 = fieldInfos.indexOf("MediaFilePaths");
//            int tag5 = fieldInfos.indexOf("Description");
//            int tag6 = fieldInfos.indexOf("HttpAddress");
//
//            if(tag1==-1 || tag2==-1 || tag3==-1 || tag4==-1 || tag5==-1) return false;
//            if(fieldInfos.get(tag1).getType() != FieldType.TEXT) return false;
//            if(fieldInfos.get(tag2).getType() != FieldType.INT16) return false;
//            if(fieldInfos.get(tag3).getType()!= FieldType.TEXT) return false;
//            if(fieldInfos.get(tag4).getType()!= FieldType.TEXT) return false;
//            if(fieldInfos.get(tag5).getType()!= FieldType.TEXT) return false;
//            if(fieldInfos.get(tag6).getType()!= FieldType.TEXT) return false;

            if (fieldInfos.indexOf("MediaFileName") == -1) {
                fieldInfo = new FieldInfo();
                fieldInfo.setType(FieldType.TEXT);
                fieldInfo.setName("MediaFileName");
                fieldInfos.add(fieldInfo);
                fieldInfo.dispose();
            }
//            if (fieldInfos.indexOf("MediaFileType") == -1) {
//                fieldInfo = new FieldInfo();
//                fieldInfo.setType(FieldType.INT16);
//                fieldInfo.setName("MediaFileType");
//                fieldInfos.add(fieldInfo);
//                fieldInfo.dispose();
//            }
            if (fieldInfos.indexOf("ModifiedDate") == -1) {
                fieldInfo = new FieldInfo();
                fieldInfo.setType(FieldType.TEXT);
                fieldInfo.setName("ModifiedDate");
                fieldInfos.add(fieldInfo);
                fieldInfo.dispose();
            }
            if (fieldInfos.indexOf("Description") == -1) {
                fieldInfo = new FieldInfo();
                fieldInfo.setType(FieldType.TEXT);
                fieldInfo.setName("Description");
                fieldInfos.add(fieldInfo);
                fieldInfo.dispose();
            }
            if (fieldInfos.indexOf("MediaFilePaths") == -1) {
                fieldInfo = new FieldInfo();
                fieldInfo.setType(FieldType.TEXT);
                fieldInfo.setName("MediaFilePaths");
                fieldInfo.setMaxLength(800);
                fieldInfos.add(fieldInfo);
                fieldInfo.dispose();
            }
            if (fieldInfos.indexOf("HttpAddress") == -1) {
                fieldInfo = new FieldInfo();
                fieldInfo.setType(FieldType.TEXT);
                fieldInfo.setName("HttpAddress");
                fieldInfos.add(fieldInfo);
                fieldInfo.dispose();
            }


//            if(pDatasetVector.getPrjCoordSys().getType() != PrjCoordSysType.PCS_EARTH_LONGITUDE_LATITUDE) return false;
        } else {
            // Dataset不存在
            try {
                DatasetVectorInfo info = new DatasetVectorInfo(datasetName, DatasetType.CAD);
                pDatasetVector = datasource.getDatasets().create(info);
            } catch (Exception e) {
                return false;
            }

            if (pDatasetVector == null) {
                return false;
            }
            FieldInfos fieldInfos = pDatasetVector.getFieldInfos();

            fieldInfo = new FieldInfo();
            fieldInfo.setType(FieldType.TEXT);
            fieldInfo.setName("MediaFileName");
            fieldInfos.add(fieldInfo);
            fieldInfo.dispose();

//            fieldInfo = new FieldInfo();
//            fieldInfo.setType(FieldType.INT16);
//            fieldInfo.setName("MediaFileType");
//            fieldInfos.add(fieldInfo);
//            fieldInfo.dispose();

            fieldInfo = new FieldInfo();
            fieldInfo.setType(FieldType.TEXT);
            fieldInfo.setName("ModifiedDate");
            fieldInfos.add(fieldInfo);
            fieldInfo.dispose();

            fieldInfo = new FieldInfo();
            fieldInfo.setType(FieldType.TEXT);
            fieldInfo.setName("Description");
            fieldInfos.add(fieldInfo);
            fieldInfo.dispose();

            fieldInfo = new FieldInfo();
            fieldInfo.setType(FieldType.TEXT);
            fieldInfo.setName("MediaFilePaths");
            fieldInfo.setMaxLength(600);
            fieldInfos.add(fieldInfo);
            fieldInfo.dispose();

            fieldInfo = new FieldInfo();
            fieldInfo.setType(FieldType.TEXT);
            fieldInfo.setName("HttpAddress");
            fieldInfos.add(fieldInfo);
            fieldInfo.dispose();

            pDatasetVector.setPrjCoordSys(new PrjCoordSys(PrjCoordSysType.PCS_EARTH_LONGITUDE_LATITUDE));
        }

        this.dataset = pDatasetVector;
        return true;
    }

    public void saveLocationDataToDataset() {
        GeoPoint point = new GeoPoint();

        Recordset recordset = ((DatasetVector)this.dataset).getRecordset(false, CursorType.DYNAMIC);
        Point2D pt = new Point2D(this.location);
        if (!SMap.safeGetType(SMap.getInstance().getSmMapWC().getMapControl().getMap().getPrjCoordSys(),PrjCoordSysType.PCS_EARTH_LONGITUDE_LATITUDE)) {
            Point2Ds point2Ds = new Point2Ds();
            point2Ds.add(pt);
            PrjCoordSys prjCoordSys = new PrjCoordSys();
            prjCoordSys.setType(PrjCoordSysType.PCS_EARTH_LONGITUDE_LATITUDE);
            CoordSysTransParameter parameter = new CoordSysTransParameter();

            CoordSysTranslator.convert(point2Ds, prjCoordSys, SMap.getInstance().getSmMapWC().getMapControl().getMap().getPrjCoordSys(), parameter, CoordSysTransMethod.MTH_GEOCENTRIC_TRANSLATION);
            pt = point2Ds.getItem(0);
        }

        point.setX(pt.getX());
        point.setY(pt.getY());
        boolean result = recordset.addNew(point);

        recordset.edit();
        recordset.moveLast();
        recordset.setString("MediaFileName", this.fileName);

        recordset.update();
        recordset.dispose();
    }

    public boolean saveMedia(Context context, ArrayList<String> filePaths, String toDictionary, boolean addNew) {
        String sdcard = android.os.Environment.getExternalStorageDirectory().getAbsolutePath();

        ArrayList<String> pathArr;
        ArrayList<String> tempFilePaths = new ArrayList<>();
        for (int i = 0; i < filePaths.size(); i++) {
            if (filePaths.get(i).indexOf("content://") == 0) {
                String  myImageUrl = filePaths.get(i);
                Uri uri = Uri.parse(myImageUrl);
                String[] proj = { MediaStore.Images.Media.DATA, MediaStore.Video.Media.DATA };
                Cursor cursor = context.getContentResolver().query(uri, proj,null,null,null);
                int actualMediaColumnIndex = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
                cursor.moveToFirst();
                String mediaPath = cursor.getString(actualMediaColumnIndex);
                tempFilePaths.add(mediaPath);
            } else {
                tempFilePaths.add(filePaths.get(i));
            }
        }
        pathArr = SMFileUtil.copyFiles(tempFilePaths, toDictionary);

        for (int i = 0; i < pathArr.size(); i++) {
            if (pathArr.get(i).indexOf(sdcard) == 0) {
                pathArr.set(i, pathArr.get(i).replace(sdcard, ""));
            }
        }

        paths = pathArr;

        if (pathArr != null && addNew) saveLocationDataToDataset();

        return pathArr != null && pathArr.size() > 0;
    }

    public boolean addMediaFiles(ArrayList<String> files) {
        if (files == null || files.size() == 0) return false;
        ArrayList<String> addingArr = new ArrayList<>();

        for (int i = 0; i < files.size(); i++) {
            boolean exsit = false;
            if (!new File(files.get(i)).exists()) continue;
            for (int j = 0; j < this.paths.size(); j++) {
                if (files.get(i).equals(this.paths.get(j))) {
                    exsit = true;
                    break;
                }
            }
            if (!exsit) {
                addingArr.add(files.get(i));
            }
        }

        this.paths.addAll(addingArr);

        return true;
    }

    public boolean deleteMediaFiles(ArrayList<String> files) {
        if (files == null || files.size() == 0) return false;
        ArrayList<String> addingArr = new ArrayList<>();

        for (int i = 0; i < files.size(); i++) {
            if (!new File(files.get(i)).exists()) continue;
            for (int j = 0; j < this.paths.size(); j++) {
                if (files.get(i).equals(this.paths.get(j)) && new File(files.get(i)).delete()) {
                    this.paths.remove(j);
                    break;
                }
            }
        }

        return true;
    }
}
